test/metabase/api/card_test.clj

```diff
@@ -4192,3 +4193,72 @@
             :moderation_status nil}
            (-> (mt/user-http-request :rasta :get 200 (str "card/" card-id))
                :dashboard)))))
+
+(deftest cannot-join-question-with-itself
+  (doseq [card-type [:question :metric :model]]
+    (testing (str "Cannot join a " card-type " with itself.")
+      (let [mp (mt/metadata-provider)
+            query (as-> (lib/query mp (lib.metadata/table mp (mt/id :orders))) $q
+                    (lib/aggregate $q (lib/count))
+                    (lib/breakout $q (m/find-first (comp #{"Created At"} :display-name)
+                                                   (lib/breakoutable-columns $q))))]
+        (mt/with-temp [:model/Card {:keys [id]} {:dataset_query (lib/->legacy-MBQL query) :type card-type}]
+          (doseq [card-type [:question :metric :model]]
+            (let [card (lib.metadata/card mp id)
+                  columns (lib/returned-columns (lib/query mp card))
+                  right-column (m/find-first (comp #{"ID"} :display-name) columns)
+                  query-with-self-join (lib/join query
+                                                 (lib/join-clause card
+                                                                  [(lib/=
+                                                                    (lib.metadata/field mp (mt/id :orders :id))
+                                                                    right-column)]))]
+              (mt/user-http-request :crowberto :put 400 (str "card/" id)
+                                  {:dataset_query (lib/->legacy-MBQL query-with-self-join)
+                                   :type card-type}))))))))
+
+(deftest cannot-save-metric-with-formula-cycle
+  (testing (str "Cannot join a metric with itself.")
+    (let [mp (mt/metadata-provider)
+          query-a (as-> (lib/query mp (lib.metadata/table mp (mt/id :orders))) $q
+                  (lib/aggregate $q (lib/count))
+                  (lib/breakout $q (m/find-first (comp #{"Created At"} :display-name)
+                                                 (lib/breakoutable-columns $q))))]
```

Comment:
Busted indentation. Also consider using `as->` only on one line for cases like this:
```clojure
  (-> (lib/query ...)
      (lib/aggregate (lib/count))
      (as-> $q (let [cols (lib/breakoutable-columns $q)]
                 (lib/breakout $q (m/find-first (comp #{"Created At"} :display-name) cols))))
```
but only where it's more readable overall.

We do this all the time with inline `cond->` too.

Reply:
:+1: